### ISSUE TITLE
`ModelOutputDict` の型安全性改善

### ISSUE THREAD

**Issue by kurorosu** [kurorosu/pochidetection#361]
## Summary

`ModelOutputDict` が `total=False` で全キーをオプショナルにしているため, RT-DETR (`pred_logits`, `pred_boxes`) と SSD (`predictions`) が同じ型を返しながら異なるキーを使用する Liskov Substitution Principle 違反が発生している.

## Background

現在 `IDetectionModel.forward()` の戻り値 `ModelOutputDict` は `TypedDict(total=False)` で定義されており, どのキーが返されるかはモデル実装に依存する. 呼び出し側は `isinstance` チェックやキー存在確認で分岐する必要があり, 型安全性が損なわれている.

## Current Implementation

```python
# pochidetection/interfaces/model.py L11-22
class ModelOutputDict(TypedDict, total=False):
    loss: torch.Tensor
    pred_logits: torch.Tensor
    pred_boxes: torch.Tensor
    predictions: list[dict[str, torch.Tensor]]
```

- RT-DETR: `{"pred_logits": ..., "pred_boxes": ...}` を返す
- SSD: `{"predictions": [...]}` を返す
- 呼び出し側で `isinstance(model, RTDetrModel)` チェックが必要

## Proposed Changes

以下のいずれかのアプローチを検討する:

1. **Generic 型パラメータ化**: `IDetectionModel[TOutput]` にして出力型をアーキテクチャごとに明示
2. **共通出力クラスの導入**: `DetectionOutput(loss, predictions: list[Detection])` のような共通フォーマットに統一
3. **アーキテクチャ別 TypedDict の分離**: `RTDetrOutputDict`, `SSDOutputDict` を定義し Union で表現

## Acceptance Criteria

- `ModelOutputDict` の型定義がアーキテクチャ間で一貫している
- 呼び出し側で `isinstance` チェックや動的キー確認が不要になる
- mypy strict モードで型エラーが発生しない
- 既存テストが全て通る

## Labels

`refactor`

## Branch

`refactor/model-output-type-safety`
